File: src/main/java/p00/Problem_0021.java
package p00;

public class Problem_0021 {

	private static final int[] SAVE = new int[10001];

	public static int solve() {

		int sum = 0;

		for (int i = 2; i < 10000; i++) {

			if (SAVE[i] == 0)
				SAVE[i] = sumOfDivisors(i);

			if (SAVE[i] < 10001) {
				if (SAVE[SAVE[i]] == 0)
					SAVE[SAVE[i]] = sumOfDivisors(SAVE[i]);

				if (i == SAVE[SAVE[i]] && i != SAVE[i])
					sum += i;

			} else if (i == sumOfDivisors(SAVE[i]) && i != SAVE[i])
				sum += i;

		}
	
		return sum;

	}

	private static int sumOfDivisors(int p) {

		if (p == 1)
			return 0;

		int sum = 1;
		double square_root = Math.sqrt(p);

		for (int i = 2; i < square_root; i++)
			if (p % i == 0)
				sum += i + (p / i);

		if (Math.pow((int) square_root, 2) == p)
			sum += square_root;

		return sum;
	}

}

File: src/main/kotlin/p00/Problem0021.kt
package p00

import kotlin.math.pow
import kotlin.math.sqrt

class Problem0021 {
    private val SAVE = IntArray(10001)
    fun solve(): Int {
        var sum = 0
        for (i in 2..9999) {
            if (SAVE[i] == 0) SAVE[i] = sumOfDivisors(i)
            if (SAVE[i] < 10001) {
                if (SAVE[SAVE[i]] == 0) SAVE[SAVE[i]] = sumOfDivisors(SAVE[i])
                if (i == SAVE[SAVE[i]] && i != SAVE[i]) sum += i
            } else if (i == sumOfDivisors(SAVE[i]) && i != SAVE[i]) sum += i
        }
        return sum
    }

    private fun sumOfDivisors(p: Int): Int {
        if (p == 1) return 0
        var sum = 1
        val squareRoot = sqrt(p.toDouble())
        var i = 2
        while (i < squareRoot) {
            if (p % i == 0) sum += i + p / i
            i++
        }
        if (squareRoot.toInt().toDouble().pow(2.0) == p.toDouble()) sum += squareRoot.toInt()
        return sum
    }
}